### ISSUE TITLE
Delay in file creation with Python implementation in ASP

### ISSUE THREAD

**Issue by ArnaudBelcour** [Aluriak/clyngor#5]
Hello,
I have an issue while using Clyngor to encapsulate ASP files (which also use python implementation). The python implementation in ASP is used to select some atoms in the answer and to write these atoms in a file. The ASP encoding and fact have been tested with Clingo (version 5.2.2 compiled with Python) and it works, the file is created.

Clyngor achieves to create the file but there is a delay in the creation of the file. Indeed the file is not accessible right after the solver call. If I try to use the file, I have an error saying that the file doesn't exist. But when the script ends, the file exists.

I write a simple example, with encoding and fact files (which are working with my Clingo) and a python script with Clyngor which creates the file but runs into an error when trying to access the file with the `print(file.read())`.

The ASP encoding extracts a number from a predicate with two numbers, assign it to another predicate and then write the new predicate in a file. 

Example:

Python script:
```python
import clyngor

clyngor.solve(('test.lp', 'data.lp'))

print(open('resultfile.lp').read())
```

Encoding file:

```
#script (python)

import clingo

def on_model(answer, output_file):
    atoms=answer.symbols(atoms=True)
    for atom in atoms:
        if atom.name == "count" and len(atom.arguments) == 1:
            output_file.write( "count("+str(atom.arguments[0])+").\n")



def main(prg):
    output_file = open("resultfile.lp", "w")
    prg.ground([("base", [])])
    for answer_number, answer in enumerate(prg.solve(yield_=True, async=True), start=1):
        print(answer)
        on_model(answer, output_file)
    output_file.close()

#end.

count(X) :- query(X, A).
```

`data.lp`:

```
query(1,2).
query(2,3).
query(4,5).
```

**Comment by ArnaudBelcour** [comment]
This is my clingo --version : 

```
clingo version 5.2.2
Address model: 64-bit

libgringo version 5.2.2
Configuration: with Python 2.7.14, without Lua

libclasp version 3.3.3 (libpotassco version 1.0.1)
Configuration: WITH_THREADS=1
Copyright (C) Benjamin Kaufmann

License: The MIT License <https://opensource.org/licenses/MIT>
```

**Comment by Aluriak** [comment]
The problem is due to `clyngor.solve` being a generator.

When, in the python file, you run solely `clyngor.solve(('test.lp', 'data.lp'))`, nothing is done. You have to get at least the first element to run the solving and therefore the side effects of the ASP source code.

One fix may be : `next(clyngor.solve(('test.lp', 'data.lp')), None)`.

**Comment by ArnaudBelcour** [comment]
Yes, I have tested this fix on my real code and it solves the problem.
Thank you!

**Comment by Aluriak** [comment]
Problem is not solved, and is probably a big one.

`encoding.lp` is:


    #script (python)

    import clingo

    def on_model(answer, output_file):
        atoms=answer.symbols(atoms=True)
        for atom in atoms:
            if atom.name == "count" and len(atom.arguments) == 1:
                output_file.write( "count("+str(atom.arguments[0])+").\n")



    def main(prg):
        output_file = open("resultfile.lp", "w")
        prg.ground([("base", [])])
        # NB: removing async=True will remove the error
        with prg.solve(yield_=True, async=True) as handle:
            for answer_number, answer in enumerate(handle, start=1):
                print(answer)
                on_model(answer, output_file)
        output_file.close()

    #end.

    count(X) :- query(X, A).
    query(1,2).
    query(2,3).
    query(4,5).

`test.py` is:

    import clyngor

    next(clyngor.solve(('test.lp',)), None)  # we need this to work
    # clyngor.solve(('test.lp', 'testt.lp'))  # but this should work too

    print(open('resultfile.lp').read())
